src/main/java/com/google/devtools/build/lib/exec/SpawnLogReconstructor.java

```diff
@@ -212,6 +225,136 @@ private static File reconstructSymlink(ExecLogEntry.UnresolvedSymlink entry) {
+      for (var file : reconstructRunfilesSymlinkTarget(newPath, symlink.getValue())) {
+        builder.put(newPath, file);
+      }
+    }
+    Collection<File> inputs = reconstructInputs(runfilesTree.getArtifactsId()).values();
+    inputs.stream()
+        .flatMap(
+            file ->
+                getRunfilesPaths(file.getPath(), runfilesTree.getLegacyExternalRunfiles())
+                    .map(
+                        relativePath ->
+                            file.toBuilder()
+                                .setPath(runfilesTree.getPath() + "/" + relativePath)
+                                .build()))
+        .forEach(file -> builder.put(file.getPath(), file));
+    getDotRunfileFileIfNeeded(runfilesTree).ifPresent(file -> builder.put(file.getPath(), file));
+    return Pair.of(runfilesTree.getPath(), ImmutableList.copyOf(builder.values()));
+  }
+
+  private Stream<String> getRunfilesPaths(String originalPath, boolean legacyExternalRunfiles) {
+    String path = originalPath;
+    if (path.startsWith("bazel-out/") || path.startsWith("blaze-out/")) {
+      // Trim the first three segments ("bazel-out/<config>/bin/") from the path.
+      int thirdSlash = path.indexOf('/', path.indexOf('/', "bazel-out/".length()) + 1);
+      path = path.substring(thirdSlash + 1);
+    }
+    if (path.startsWith(LabelConstants.EXTERNAL_PATH_PREFIX.getPathString())) {
+      Stream.Builder<String> paths = Stream.builder();
+      paths.add(path.substring(LabelConstants.EXTERNAL_PATH_PREFIX.getPathString().length()));
+      if (legacyExternalRunfiles) {
+        paths.add(workspaceRunfilesDirectory + "/" + path);
+      }
+      return paths.build();
+    }
+    return Stream.of(workspaceRunfilesDirectory + "/" + path);
+  }
+
+  private Collection<File> reconstructRunfilesSymlinkTarget(
+      String newPath, ExecLogEntry.RunfilesTree.SymlinkTarget target) throws IOException {
+    return switch (target.getTargetCase()) {
+      case EMPTY_FILE -> ImmutableList.of(File.newBuilder().setPath(newPath).build());
+      case FILE_ID -> {
+        var file = getFromMap(fileMap, target.getFileId());
+        yield ImmutableList.of(file.toBuilder().setPath(newPath).build());
+      }
+      case UNRESOLVED_SYMLINK_ID -> {
+        var symlink = getFromMap(symlinkMap, target.getUnresolvedSymlinkId());
+        yield ImmutableList.of(symlink.toBuilder().setPath(newPath).build());
+      }
+      case DIRECTORY_ID -> {
+        var dir = getFromMap(dirMap, target.getDirectoryId());
+        yield dir.getSecond().stream()
+            .map(
+                file ->
+                    file.toBuilder()
+                        .setPath(newPath + file.getPath().substring(dir.getFirst().length()))
+                        .build())
+            .collect(toImmutableList());
+      }
+      default ->
+          throw new IOException(
+              String.format("unknown target type %d", target.getTargetCase().getNumber()));
```

Comment:
Add an explicit case for `RUNFILES_TREE_ID` with a better error message. (Useful to anyone touching the compact log implementation in the future and using the reconstructor to verify that the changes are correct.)

Reply:
Yes, that's what I meant, sorry.